### PR TITLE
overriding input default classes with custom tailwind classes

### PR DESCRIPTION
**Pull request by ShivanshPlays**
---
name: Pull request
about: Maintain this format while making a PR
title: ""
labels: ""
assignees: "shivansh"
---

## Description
I have modified the code to override the className provided by the user over default css. 
The user just has to make 1 change in the tailwind.config.js of their project by adding [important: "#root"] in the default export and they are good to go....now they can add tailwind classes and the changes will reflect.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant:

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update (changes to documentation only)
- [ ] Chore (cleanup or minor improvement that does not change functionality)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings or errors.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.

## Screenshots (if applicable)
image
image
image

If applicable, please add screenshots to illustrate the changes.

## Additional Information

Please add any other relevant information or context for the reviewers.

### CODE REVIEW

File: package.json
Comment (Git21221) [inline package.json:1]: why clsx and tw-merge is in dependencies not devDependencies?

File: src/components/Input/Input.tsx
Comment (Git21221) [inline src/components/Input/Input.tsx:1]: className is optional field, make sure to put it in @param also (JSDoc)

File: test/tailwind.config.js
Comment (Git21221) [inline test/tailwind.config.js:1]: why `important: "#root"` in tailwind config?

**Comment by Git21221** [comment]
@ShivanshPlays  PR should be on `dev` branch not `main`. main is connected  to NPM and github packages.

**Comment by ShivanshPlays** [comment]
Apologies, I have updated the branch.

**Comment by Git21221** [comment]
@ShivanshPlays follow the reviews comments

**Comment by ShivanshPlays** [comment]
Sorry for the confusion, this is my first PR, can you please elaborate what does "follow the reviews comments" mean....if youhave commented somwhere, I can't find it

**Comment by Git21221** [comment]
@ShivanshPlays  see above, 
image
I have added my review

**Comment by ShivanshPlays** [comment]
Screenshot 2024-10-02 154630
I cant find your comments to reply on them....

1. If tw-merge and clsx are required for your package to function, meaning that the consumers of your npm package will rely on these utilities when they import your package, then these libraries should be in dependencies.

2. okay I'm on it

3. The root is important for users because by default tailwind classes conflicting with mui gets lesser priority and the changes dont reflect properly, so adding important=#root gives user's className more priority.

**Comment by ShivanshPlays** [comment]
Please check @Git21221

**Comment by Git21221** [comment]
@ShivanshPlays 
1) if user install our package everything will be installed, so if we include it in devDependencies it will also work
3) is it possible not to change anything from user side, only they will install it and start working on it, no need to change any config file? is it possible?

**Comment by ShivanshPlays** [comment]
I was wondering the exact same thing...actually, what happened is that yesterday I was trying to override the default CSS, but it was not working. 
Then after spending yesterday's whole day, I realise that this was due to tailwind loading its CSS before anything else. So the only work around, I could find was this one. 

As for the dev dependencies, I have updated it. Please check.

**Comment by Git21221** [comment]
@ShivanshPlays will it create any other issue? like we know inline css has the highest priority so, if someone changes the config as you have done in test and then try to override tailwind with inline css, will it have any issue?
if not we can go with changing the config file for now, if we found any better approach, we can shift

**Comment by Git21221** [comment]
@ShivanshPlays any progress?

**Comment by ShivanshPlays** [comment]
yeah, sorry for the delay, I'm on it

File: package.json
Comment (Git21221) [reply package.json:1]: it should be a devdependency

File: test/tailwind.config.js
Comment (ShivanshPlays) [reply test/tailwind.config.js:1]: tailwind loading its CSS before anything else. So the only work around I could find was important: "#root"

File: src/components/Input/Input.tsx
Comment (ShivanshPlays) [reply src/components/Input/Input.tsx:1]: done

File: package.json
Comment (ShivanshPlays) [reply package.json:1]: done

**Comment by ShivanshPlays** [comment]
Hi @Git21221 , you forgot the tags gssoc gssoc-ext

**Comment by Git21221** [comment]
@ShivanshPlays let me know if it's reflected in leaderboard

**Comment by ShivanshPlays** [comment]
ok thanku

**Pull request merged**